### DISCUSSION TITLE
💅 useFilenamingConvention ignores style in extension

### DISCUSSION THREAD

**Discussion by kyptov** [biomejs/biome#3650] [Ideas]
### Environment information

```bash
D:\tms-api git:[feat-highway-api]
biome rage --linter
CLI:
  Version:                      1.8.3
  Color support:                true

Platform:
  CPU Architecture:             x86_64
  OS:                           windows

Environment:
  BIOME_LOG_DIR:                unset
  NO_COLOR:                     unset
  TERM:                         unset
  JS_RUNTIME_VERSION:           "v20.16.0"
  JS_RUNTIME_NAME:              "node"
  NODE_PACKAGE_MANAGER:         unset

Biome Configuration:
  Status:                       Loaded successfully
  Formatter disabled:           false
  Linter disabled:              false
  Organize imports disabled:    false
  VCS disabled:                 true

Linter:
  JavaScript enabled:           true
  JSON enabled:                 true
  CSS enabled:                  false
  Recommended:                  false
  All:                          false
  Enabled rules:
  complexity/noBannedTypes
  suspicious/noCatchAssign
  suspicious/noPrototypeBuiltins
  complexity/useLiteralKeys
  suspicious/noMisleadingInstantiator
  suspicious/noDebugger
  complexity/noMultipleSpacesInRegularExpressionLiterals
  complexity/noUselessLoneBlockStatements
  complexity/noVoid
  suspicious/noGlobalAssign
  suspicious/noUnsafeNegation
  suspicious/useValidTypeof
  correctness/noConstantCondition
  suspicious/noDuplicateParameters
  correctness/noEmptyPattern
  suspicious/noExtraNonNullAssertion
  complexity/noUselessTernary
  style/useConst
  suspicious/noCompareNegZero
  suspicious/noSelfCompare
  correctness/noUnreachableSuper
  security/noGlobalEval
  style/noNonNullAssertion
  correctness/noConstAssign
  correctness/noSwitchDeclarations
  correctness/noUnusedVariables
  suspicious/noAsyncPromiseExecutor
  correctness/useIsNan
  suspicious/noControlCharactersInRegex
  complexity/noUselessTypeConstraint
  style/noVar
  suspicious/noDoubleEquals
  style/useConsistentArrayType
  complexity/noWith
  suspicious/noDuplicateClassMembers
  complexity/noExtraBooleanCast
  suspicious/noMisleadingCharacterClass
  correctness/noPrecisionLoss
  suspicious/noRedeclare
  correctness/noInvalidConstructorSuper
  suspicious/noDuplicateObjectKeys
  suspicious/noFallthroughSwitchClause
  correctness/noUnreachable
  suspicious/noConfusingVoidType
  complexity/useOptionalChain
  suspicious/noDuplicateCase
  style/useFilenamingConvention
  complexity/useRegexLiterals
  correctness/noSelfAssign
  style/useShorthandFunctionType
  suspicious/noEmptyBlockStatements
  suspicious/noShadowRestrictedNames
  complexity/noUselessCatch
  correctness/noInvalidUseBeforeDeclaration
  correctness/noUnsafeFinally
  style/useBlockStatements
  complexity/noUselessRename
  complexity/noUselessConstructor
  style/useEnumInitializers
  style/noDefaultExport
  style/useSingleVarDeclarator
  suspicious/noImportAssign
  style/noCommaOperator
  suspicious/useDefaultSwitchClauseLast
  style/noNamespace
  correctness/noGlobalObjectCalls
  correctness/noNewSymbol
  correctness/noEmptyCharacterClassInRegex
  suspicious/noClassAssign
  suspicious/noFunctionAssign

Workspace:
  Open Documents:               0
```

### Rule name

useFilenamingConvention

### Playground link

Not reproducible in the playground.

### Expected result

In our project, we are using kebab case for file naming. To distinguish between "layers," we are using suffixes. 
For User, we might have several files: 
`user.entity.ts, user.service.ts, user.repository.ts, user.use-case.ts.` This works really well. 

However, the problem is with `user.use-case.ts` Biome shows an error: `The file extension should be in lowercase without any special characters.`
Other examples include `user.unit-test.ts` and `user.e2e-test.ts`. 

**The expected result is no error for the second and following extensions.**

### Code of Conduct

- [X] I agree to follow Biome's Code of Conduct

**Comment by ematipico** [comment]
Converted to a discussion -> Idea because that's not a bug, as Biome works as expected.

Please share a possible solution to improve the rule, because it's the first that hear about "style in extensions"

**Reply by kyptov** [reply]
It is commonly advisable to name tests with extensions like `*.test.ts` or `*.spec.ts`. This allows us to use the second extension to give files additional meaning. If it's one word, there is no problem, but what if it is more complex, such as Unit Test or E2E Test? Options like `*.unit.test.ts`, `*.unittest.ts`, or `*.UnitTest.ts` come to mind. It appears that this can be done in various ways, so it is better to select one variant and stick to it. Which one should we choose? Let's choose the one we are already using for file naming. 

A possible solution is to treat all extensions except the last as part of the filename and apply the same rules.

**Reply by Conaclos** [reply]
We accept a case named `export` that requires to use the name of an export. If a user chooses to enforce `export` for filenames, this doesn't make sense of applying it to extensions.

In that case, should we provide another config for "subextensions" (I don't know if there is a better name to designate them)?

**Reply by kyptov** [reply]
> In that case, should we provide another config for "subextensions" (I don't know if there is a better name to designate them)?

I suggest not to add new extension. It can be config "extensionAfterLastDot: true" default to false (current behavior). So it will be always just filename and extension. If true - first dot is delimiters.

**Reply by Conaclos** [reply]
Another approach could be adding a config with a regex based approach. We could have two capture groups, one that extracts the filename and the other that extracts the extension (possibly with subextensions). This could also allow users to support arbitrary filenames.

**Reply by quezak** [reply]
> because it's the first that hear about "style in extensions"

To me it's the opposite, I didn't expect the current rule's behavior because I always considered only the last dot-delimited suffix the "extension", and the middle ones more like "file categories". In our case also the middle extensions can contain dashes, e.g. we use `*.type-spec.ts` for type-level tests, but also a multitude of other "file categories" like the already mentioned `*.use-case.ts`.

---

IMO a good solution would be to add a configuration option to the rule to disallow dashes only in the last extension.

**Reply by Conaclos** [reply]
I have worked on a fix that applies the configured format (excluding `export` and adding `lowercase`) to all extensions.
Thus, with the default configuration, the following names will pass: `filename.dashed-ext.dashed-ext`, `filename.underscore_ext.underscore_ext` and `filename.camelExt.camelExt`.

**Reply by Conaclos** [reply]
Fixed in #3958
